File: frontend/src/types/job.ts
export interface JobSkill {
  id: string;
  skill_id: string;
  name: string;
  category?: string | null;
  is_required: boolean;
  importance_weight: number;
  min_proficiency: string;
}

export interface JobItem {
  id: string;
  title: string;
  company: string;
  location: string;
  employment_type: string;
  experience_level: string;
  min_experience_years: number;
  target_education_level: string;
  description: string;
  salary_range?: string | null;
  is_active: boolean;
  created_at: string;
  skills: JobSkill[];
  is_saved: boolean;
  match_score?: number | null;
}

export interface EducationCompatibility {
  candidate_highest_level: string;
  required_level: string;
  meets_requirement: boolean;
  score: number;
  explanation: string;
}

export interface MatchedSkillDetail {
  name: string;
  candidate_proficiency: string;
  required_proficiency: string;
  importance_weight: number;
  credit: number;
  is_exact: boolean;
  confidence: number;
  is_required: boolean;
}

export interface PartialSkillDetail {
  job_skill_name: string;
  candidate_skill_name: string;
  similarity_weight: number;
  credit: number;
  candidate_proficiency: string;
  required_proficiency: string;
  is_required: boolean;
}

export interface MissingSkillDetail {
  name: string;
  required_proficiency: string;
  importance_weight: number;
  is_required: boolean;
}

export interface JobMatchBreakdown {
  overall_score: number;
  required_skills_score: number;
  preferred_skills_score: number | null;
  experience_score: number;
  education_score: number;
  matched_skills: MatchedSkillDetail[];
  partial_skills: PartialSkillDetail[];
  missing_required_skills: MissingSkillDetail[];
  missing_preferred_skills: MissingSkillDetail[];
  experience_gap: number;
  experience_gap_text: string;
  education_compatibility: {
    candidate_highest_level: string;
    required_level: string;
    meets_requirement: boolean;
    score: number;
    explanation: string;
  };
  total_required_skills_count: number;
  matched_required_skills_count: number;
  total_preferred_skills_count: number;
  matched_preferred_skills_count: number;
  active_weights: Record<string, number>;
  explanation: string;
}

export interface JobDetail extends JobItem {
  cached_match?: JobMatchBreakdown | null;
}

export interface PaginatedJobsResponse {
  items: JobItem[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface SavedJobItem {
  id: string;
  job_id: string;
  created_at: string;
  job: JobItem;
}

export interface JobFilterParams {
  search?: string;
  role?: string;
  location?: string;
  employment_type?: string;
  experience_level?: string;
  skill?: string;
  saved_only?: boolean;
  page?: number;
  limit?: number;
}

File: backend/app/schemas/job.py
"""
Pydantic Schemas for Job Postings, Saved Jobs, Discovery Filters, and Match Evaluation.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field


class JobSkillResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    skill_id: UUID
    name: str
    category: Optional[str] = None
    is_required: bool
    importance_weight: float
    min_proficiency: str


class JobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    company: str
    location: str
    employment_type: str
    experience_level: str
    min_experience_years: float
    target_education_level: str
    description: str
    salary_range: Optional[str] = None
    is_active: bool
    created_at: datetime
    skills: List[JobSkillResponse] = Field(default_factory=list)
    is_saved: bool = False
    match_score: Optional[float] = None


class JobDetailResponse(JobResponse):
    cached_match: Optional[Dict[str, Any]] = None


class PaginatedJobsResponse(BaseModel):
    items: List[JobResponse]
    total: int
    page: int
    limit: int
    pages: int


class SavedJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    job_id: UUID
    created_at: datetime
    job: JobResponse


class EducationCompatibilityResponse(BaseModel):
    candidate_highest_level: str
    required_level: str
    meets_requirement: bool
    score: float
    explanation: str


class JobMatchBreakdownResponse(BaseModel):
    overall_score: float = Field(..., ge=0.0, le=100.0)
    required_skills_score: float = Field(..., ge=0.0, le=100.0)
    preferred_skills_score: Optional[float] = Field(None, ge=0.0, le=100.0)
    experience_score: float = Field(..., ge=0.0, le=100.0)
    education_score: float = Field(..., ge=0.0, le=100.0)
    matched_skills: List[Dict[str, Any]] = Field(default_factory=list)
    partial_skills: List[Dict[str, Any]] = Field(default_factory=list)
    missing_required_skills: List[Dict[str, Any]] = Field(default_factory=list)
    missing_preferred_skills: List[Dict[str, Any]] = Field(default_factory=list)
    experience_gap: float
    experience_gap_text: str
    education_compatibility: EducationCompatibilityResponse
    total_required_skills_count: int
    matched_required_skills_count: int
    total_preferred_skills_count: int
    matched_preferred_skills_count: int
    active_weights: Dict[str, float]
    explanation: str
